src/lib/CofThy.ml

```diff
@@ -1,209 +1,233 @@
     match thy with
-    | `Inconsistent -> zero
-    | `Consistent thy -> left_invert' ~zero ~seq ~fast_track thy phis cont
+    | `Inconsistent -> []
+    | `Consistent thy' ->
+      match disect_cofibrations cofs with
+      | [] -> []
+      | [vars, []] when VarSet.is_empty vars -> [`Consistent thy']
+      | disected_cofs ->
+        let cached_branches =
+          drop_useless_branches @@
+          shrink_branches thy' disected_cofs in
+        List.map (fun (thy', _) -> `Consistent thy') cached_branches
+
+  let left_invert_under_cofs ~zero ~seq (thy : t) cofs cont =
+    match split thy cofs with
+    | [] -> zero
+    | [thy] -> cont thy
+    | thys -> seq cont thys
 end
 
-
-(* Invariant: local.classes must be consistent. *)
-let rec test (local : alg_thy') : cof -> bool =
-  function
-  | Cof.Cof phi ->
-    begin
-      match phi with
-      | Cof.Eq (r, s) when r = s ->
-        true
-      | Cof.Eq (r, s) ->
-        find_class local.classes r = find_class local.classes s
-      | Cof.Join phis ->
-        List.exists (test local) phis
-      | Cof.Meet phis ->
-        List.for_all (test local) phis
-    end
-  | Cof.Var v ->
-    VarSet.mem v local.true_vars
-
-let test_sequent thy cx phi =
-  let fast_track x y = if x () then true else y () in
-  Search.left_invert ~zero:true ~seq:List.for_all ~fast_track thy cx @@ fun thy ->
-  test thy phi
-
-let assume thy phis =
-  match thy with
-  | `Inconsistent -> `Inconsistent
-  | `Consistent thy ->
-    match SearchHelper.push_cofs thy @@ List.map Cof.reduce phis (* do we want Cof.reduce here? *) with
-    | None -> `Inconsistent
-    | Some thy ->
-      if SearchHelper.is_consistent thy
-      then `Consistent thy
-      else `Inconsistent
-
-let left_invert_under_cofs ~zero ~seq thy phis cont =
-  Search.left_invert ~zero ~seq ~fast_track:(fun _ x -> x ()) thy phis @@ fun alg_thy' ->
-  cont @@ `Consistent (disj_thy'_of_alg_thy' alg_thy')
-
-
-module Algebraic =
+module Disj =
 struct
-  let init () =
-    `Consistent
-      {classes = UF.init ~size:100;
-       true_vars = VarSet.empty}
+  type t = disj_thy
 
-  let consistency =
-    function
-    | `Consistent _ -> `Consistent
-    | `Inconsistent -> `Inconsistent
-
-  let disj_envelope = disj_thy_of_alg_thy
+  let init : t = [Alg.init', (VarSet.empty, [])]
 
-  let partition_thy =
+  let consistency =
     function
-    | `Inconsistent -> `Inconsistent, []
-    | `Consistent {classes; true_vars; irreducible_joins} ->
-      `Consistent {classes; true_vars}, List.map Cof.join irreducible_joins
-
-  let assemble_thy alg_thy =
-    assume @@ disj_envelope alg_thy
-
-  let left_invert_under_cofs ~zero ~seq alg_thy phis cont =
-    Search.left_invert ~zero ~seq ~fast_track:(fun _ x -> x ()) (disj_envelope alg_thy) phis @@ fun alg_thy' ->
-    cont @@ `Consistent alg_thy'
+    | [] -> `Inconsistent
+    | _ -> `Consistent
+
+  (* favonia: this optimization seems to be too costly? *)
+  let rebase_branch cached_branches : t =
```

Comment:
So what does rebase do?

Reply:
The idea is that, if there are two branches labeled as `phi1 /\ phi2` and `phi1 /\ phi2'`, then it will try to remove the common part `phi1` from all branch labels. There was an actual rebasing but after refactoring maybe it should be called "refactor".

Reply:
oh neat. what do you think about this optimization? is it useful?

Reply:
It seems to help after switching to Map. Maybe it would also help when using the hash tables because I polished the function so much after the switch. Anyways, it is not changing the semantics, so you can always replace it with the identity function and see what would happen.

Reply:
Its benefits probably come from faster dropping of useless branches. I need the branch labels for dropping useless branches, and this function tries to factor out the parts of the labels (for faster future dropping).